pkg/controller/csr/csr_controller.go

```diff
@@ -292,21 +349,50 @@ func (r *reconcileCSR) validate(csr *certificatesv1.CertificateSigningRequest, p
 		return nil, err
 	}
 
-	// The CSRName is formatted as "<secretName>:<podName>".
+	// The CSRName is formatted as follows:
+	// - Pod: "<secretName>:<podName>"
+	// - HostEndpoint: "<secretName>:<hostname>"
 	nameChunks := strings.Split(csr.Name, ":")
-	if len(nameChunks) != 2 || nameChunks[1] != pod.Name {
+	if len(nameChunks) != 2 || nameChunks[1] != expectedName {
 		return nil, fmt.Errorf("invalid: CSR name does not match expected format: %s", csr.Name)
 	}
 	secretName := nameChunks[0]
 	// Validate whether this is a CSR we monitor at all.
-	asset, ok := r.allowedTLSAssets[secretName]
+	asset, ok := allowedTLSAssets[secretName]
 	if !ok {
 		return nil, fmt.Errorf("invalid: this controller is not configured to sign secretName: %s", secretName)
 	}
 
 	// Validate whether the requestor of the CSR is registered for the given CSR
-	if fmt.Sprintf("system:serviceaccount:%s:%s", asset.serviceaccountNamespace, asset.serviceaccountName) != csr.Spec.Username {
-		return nil, fmt.Errorf("invalid requestor %s for CSR with name %s", csr.Spec.Username, csr.Name)
+	if asset.serviceaccountNamespace != "" && asset.serviceaccountName != "" {
+		if fmt.Sprintf("system:serviceaccount:%s:%s", asset.serviceaccountNamespace, asset.serviceaccountName) != csr.Spec.Username {
+			return nil, fmt.Errorf("invalid requestor %s for CSR with name %s", csr.Spec.Username, csr.Name)
+		}
+	} else {
+		// This CSR originals from non-cluster hosts. We allow multiple service accounts for different groups
+		// of non-cluster hosts. To ensure proper access control, we need to validate the requestor's permission.
+		review := &authv1.SubjectAccessReview{
+			Spec: authv1.SubjectAccessReviewSpec{
+				User:   csr.Spec.Username,
+				Groups: csr.Spec.Groups,
+				UID:    csr.Spec.UID,
+				Extra:  convertExtraValue(csr.Spec.Extra),
+				ResourceAttributes: &authv1.ResourceAttributes{
+					Group:       "certificates.tigera.io",
+					Resource:    "certificatesigningrequests",
+					Subresource: render.TyphaCommonName + render.TyphaNonClusterHostSuffix,
+					Verb:        "create",
+					Name:        secretName,
+				},
+			},
+		}
+
+		result, err := clientset.AuthorizationV1().SubjectAccessReviews().Create(ctx, review, metav1.CreateOptions{})
```

Comment:
What is the timeout set for this context. Consider setting one, if not

Reply:
I have set a 10-second timeout for this subject access review request.